# Lab 5 : CS 6760

## Introduction to Privacy Engineering, Summer 2021


## Web Scarping with Beautiful Soup

Web scraping is typically defined as a use of a program or algorithm to extract, and process data from the web. In this lab, you will learn:
• How to extract data from the web using Python's Beautiful Soup module,
• How to manipulate and clean data using Python's Pandas library,
• How to visualize data using Python's Matplotlib library.

##### Installation:

Same as Lab 1. <br>No additional packages are required.

##### This lab contains 1 task, listed below.

##### Grading rubric: for each question,
100% of the points - Correct code, correct output<br>
50% of the points - Minor logical error, partial output<br>
0% of the points - No attempt, incomplete code, wrong output

##### Submission instructions

You will have to submit the completed jupyter notebook file (.ipynb) in Canvas. <br> 
Please rename the submission file in the 'FirstName-LastName-Lab5.ipynb'format.

##### Acknowledgement

This lab assignment was modified from the Sicelo Masango's tutorial, "Web Scarping Using Python"
https://www.datacamp.com/community/tutorials/web-scraping-using-python


The datasets used in this tutorial include:
- The 10K race that took place in Hillsboro, OR on June 2017, available: http://www.hubertiming.com/results/2017GPTR10
- The Olympic medal count by country, available here: https://worldpopulationreview.com/country-rankings/olympic-medals-by-country

Importing libraries and functions

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

To perform web scraping, we also need to import the libraries shown below. The urllib.request module is used to open URLs. The Beautiful Soup package is used to extract data from html files. The Beautiful Soup library's name is bs4 which stands for Beautiful Soup, version 4.


In [ ]:
from urllib.request import urlopen
from bs4 import BeautifulSoup

After importing the necessary modules, we also need specify the URL containing the dataset, and pass it to urlopen() to get the html of the page:

In [ ]:
url = "http://www.hubertiming.com/results/2017GPTR10K"
html = urlopen(url)

Getting the html of the page is just the first step. Next step is to create a Beautiful Soup object from the html. This is done by passing the html to the BeautifulSoup() function. The Beautiful Soup package is used to parse the html, that is, take the raw html text and break it into Python objects. The second argument 'lxml' is the html parser.

In [ ]:
soup = BeautifulSoup(html, 'lxml')
type(soup)

bs4.BeautifulSoup

The soup object allows us to extract interesting information about the website we're scraping, such as getting the title of the page as shown below.

In [ ]:
# Get the title
title = soup.title
print(title)

<title>Race results for the 2017 Intel Great Place to Run \ Urban Clash Games!</title>


We can also get the text of the webpage, and quickly print it out to check if it is what we expect.

In [ ]:
# Print out the text
text = soup.get_text()
#print(soup.text)


Next, we can use the find_all() method of soup to extract useful html tags within a webpage. Examples of useful tags include < a > for hyperlinks, < table > for tables, < tr > for table rows, < th > for table headers, and < td > for table cells. The code below shows how to extract all the hyperlinks within the webpage.

In [ ]:
soup.find_all('a')


[<a href="mailto:timing@hubertiming.com">timing@hubertiming.com</a>,
 <a href="https://www.hubertiming.com/">Huber Timing Home</a>,
 <a class="btn btn-primary btn-lg" href="/results/2017GPTR#individual" role="button" style="margin: 0px 0px 5px 5px"><i aria-hidden="true" class="fa fa-user"></i> 5K Individual</a>,
 <a class="btn btn-primary btn-lg" href="/results/team/2017GPTR" role="button" style="margin: 0px 0px 5px 5px"><i aria-hidden="true" class="fa fa-users"></i> 5K Team</a>,
 <a class="btn btn-primary btn-lg" href="/results/team/2017GPTR10K" role="button" style="margin: 0px 0px 5px 5px"><i aria-hidden="true" class="fa fa-users"></i> 10K Team</a>,
 <a class="btn btn-primary btn-lg" href="/results/summary/2017GPTR10K" role="button" style="margin: 0px 0px 5px 5px"><i class="fa fa-stream"></i> Summary</a>,
 <a id="individual" name="individual"></a>,
 <a data-url="/results/2017GPTR10K" href="#tabs-1" id="rootTab" style="font-size: 18px">10K Results</a>,
 <a href="https://www.hubertimin

As we can see from the output above, html tags sometimes come with attributes such as class, src, etc. These attributes provide additional information about html elements. We can use a for loop and the get('"href") method to extract and print out only hyperlinks.

In [ ]:
all_links = soup.find_all("a")
for link in all_links:
    print(link.get("href"))

mailto:timing@hubertiming.com
https://www.hubertiming.com/
/results/2017GPTR#individual
/results/team/2017GPTR
/results/team/2017GPTR10K
/results/summary/2017GPTR10K
None
#tabs-1
https://www.hubertiming.com/
https://facebook.com/hubertiming/
None


To print out table rows only, we pass the 'tr' argument in soup.find_all().

In [ ]:
# Print the first 10 rows for sanity check
rows = soup.find_all('tr')
print(rows[:10])

[<tr colspan="2">
<b>10K:</b>
</tr>, <tr>
<td>Finishers:</td>
<td>577</td>
</tr>, <tr>
<td>Male:</td>
<td>414</td>
</tr>, <tr>
<td>Female:</td>
<td>163</td>
</tr>, <tr class="header">
<th>Place</th>
<th>Bib</th>
<th>Name</th>
<th>Gender</th>
<th>City</th>
<th>State</th>
<th>Chip Time</th>
<th>Chip Pace</th>
<th>Gun Time</th>
<th>Team</th>
</tr>, <tr data-bib="814">
<td>1</td>
<td>814</td>
<td>

                    JARED WILSON

                </td>
<td>M</td>
<td>TIGARD</td>
<td>OR</td>
<td>36:21</td>
<td>5:51</td>
<td>36:24</td>
<td></td>
</tr>, <tr data-bib="573">
<td>2</td>
<td>573</td>
<td>

                    NATHAN A SUSTERSIC

                </td>
<td>M</td>
<td>PORTLAND</td>
<td>OR</td>
<td>36:42</td>
<td>5:55</td>
<td>36:45</td>
<td>
<img class="lazy teamThumbs" data-src="/teamLogoThumbnail/logo?teamName=INTEL%20TEAM%20F&amp;raceId=1251&amp;state=OR"/>
                            INTEL TEAM F
                        </td>
</tr>, <tr data-bib="687">
<td>3</td>
<td>687</td>
<

Our next goal is take a table from a webpage, and convert it into a dataframe for easier manipulation using Python. To get there, we need to get all table rows in list form first, and then convert that list into a dataframe. Below is a for loop that iterates through table rows, and prints out the cells of the rows.

In [ ]:
for row in rows:
    row_td = row.find_all('td')
print(row_td)
type(row_td)

[<td>577</td>, <td>443</td>, <td>

                    LIBBY B MITCHELL

                </td>, <td>F</td>, <td>HILLSBORO</td>, <td>OR</td>, <td>1:41:18</td>, <td>16:20</td>, <td>1:42:10</td>, <td></td>]


bs4.element.ResultSet

The output above shows that each row is printed with html tags embedded in each row. This is not what we want. We can use remove the html tags using Beautiful Soup or regular expressions.
The easiest way to remove html tags is to use Beautiful Soup, and it takes just one line of code to do this. We pass the string of interest into BeautifulSoup(), and use the get_text() method to extract the text without html tags.

In [ ]:
str_cells = str(row_td)
cleantext = BeautifulSoup(str_cells, "lxml").get_text()
print(cleantext) 

[577, 443, 

                    LIBBY B MITCHELL

                , F, HILLSBORO, OR, 1:41:18, 16:20, 1:42:10, ]


Our next step is to convert the list into a dataframe, and get a quick view of the first 10 rows using Pandas.

In [ ]:
import re

list_rows = []
for row in rows:
    cells = row.find_all('td')
    str_cells = str(cells)
    clean = re.compile('<.*?>')
    clean2 = (re.sub(clean, '',str_cells))
    list_rows.append(clean2)
print(clean2)
type(clean2)

df = pd.DataFrame(list_rows)
df.head(10)

[577, 443, 

                    LIBBY B MITCHELL

                , F, HILLSBORO, OR, 1:41:18, 16:20, 1:42:10, ]


0
0                                                 []
1                                  [Finishers:, 577]
2                                       [Male:, 414]
3                                     [Female:, 163]
4                                                 []
5  [1, 814, \r\n\r\n                    JARED WIL...
6  [2, 573, \r\n\r\n                    NATHAN A ...
7  [3, 687, \r\n\r\n                    FRANCISCO...
8  [4, 623, \r\n\r\n                    PAUL MORR...
9  [5, 569, \r\n\r\n                    DEREK G O...

The dataframe is not in the format we want. To clean it up, we need to split the "0" column into multiple columns at the comma position. This is accomplished by using the str.split() method.

In [ ]:
df1 = df[0].str.split(',', expand=True)
df1.head(10)

0  ...                                                  9
0           []  ...                                               None
1  [Finishers:  ...                                               None
2       [Male:  ...                                               None
3     [Female:  ...                                               None
4           []  ...                                               None
5           [1  ...                                                  ]
6           [2  ...   \n\r\n                            INTEL TEAM ...
7           [3  ...                                                  ]
8           [4  ...                                                  ]
9           [5  ...   \n\r\n                            INTEL TEAM ...

[10 rows x 10 columns]

Looking at the result, we notice that the dataframe has unwanted square brackets surrounding each row. We can use the strip() method to remove the opening square bracket on column "0."

In [ ]:
df1[0] = df1[0].str.strip('[')
df1.head(10)

0      1  ...       8                                                  9
0           ]   None  ...    None                                               None
1  Finishers:   577]  ...    None                                               None
2       Male:   414]  ...    None                                               None
3     Female:   163]  ...    None                                               None
4           ]   None  ...    None                                               None
5           1    814  ...   36:24                                                  ]
6           2    573  ...   36:45   \n\r\n                            INTEL TEAM ...
7           3    687  ...   37:48                                                  ]
8           4    623  ...   38:37                                                  ]
9           5    569  ...   39:24   \n\r\n                            INTEL TEAM ...

[10 rows x 10 columns]

The table is missing table headers. We can use the find_all() method to get the table headers.

In [ ]:
col_labels = soup.find_all('th')

Similar to table rows, we can use Beautiful Soup to extract text in between html tags for table headers:

In [ ]:
all_header = []
col_str = str(col_labels)
cleantext2 = BeautifulSoup(col_str, "lxml").get_text()
all_header.append(cleantext2)
print(all_header)

['[Place, Bib, Name, Gender, City, State, Chip Time, Chip Pace, Gun Time, Team]']


We can now convert the list of headers into a pandas dataframe.

In [ ]:
df2 = pd.DataFrame(all_header)
df2.head()

0
0  [Place, Bib, Name, Gender, City, State, Chip T...

Similarly, we can split column "0" into multiple columns at the comma position for all rows.

In [ ]:
df3 = df2[0].str.split(',', expand=True)
df3.head()

0     1      2        3  ...           6           7          8       9
0  [Place   Bib   Name   Gender  ...   Chip Time   Chip Pace   Gun Time   Team]

[1 rows x 10 columns]

The two dataframes can be concatenated into one using the concat() method:

In [ ]:
frames = [df3, df1]

df4 = pd.concat(frames)
df4.head(10)

0  ...                                                  9
0      [Place  ...                                              Team]
0           ]  ...                                               None
1  Finishers:  ...                                               None
2       Male:  ...                                               None
3     Female:  ...                                               None
4           ]  ...                                               None
5           1  ...                                                  ]
6           2  ...   \n\r\n                            INTEL TEAM ...
7           3  ...                                                  ]
8           4  ...                                                  ]

[10 rows x 10 columns]

We can assign the first row to be the table header.

In [ ]:
df5 = df4.rename(columns=df4.iloc[0])
df5.head()

[Place    Bib   Name   Gender  ...   Chip Time   Chip Pace   Gun Time   Team]
0      [Place    Bib   Name   Gender  ...   Chip Time   Chip Pace   Gun Time   Team]
0           ]   None   None     None  ...        None        None       None    None
1  Finishers:   577]   None     None  ...        None        None       None    None
2       Male:   414]   None     None  ...        None        None       None    None
3     Female:   163]   None     None  ...        None        None       None    None

[5 rows x 10 columns]

Next, notice how the table header is replicated as the first row in df5. It can be dropped using the following line of code.

In [ ]:
df6 = df5.drop(df5.index[0])
df6.head()

[Place    Bib  ...  Gun Time  Team]
1  Finishers:   577]  ...      None   None
2       Male:   414]  ...      None   None
3     Female:   163]  ...      None   None
4           ]   None  ...      None   None
5           1    814  ...     36:24      ]

[5 rows x 10 columns]

In [ ]:
df6.info()
df6.shape

<class 'pandas.core.frame.DataFrame'>
Int64Index: 581 entries, 1 to 581
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   [Place      581 non-null    object
 1    Bib        580 non-null    object
 2    Name       577 non-null    object
 3    Gender     577 non-null    object
 4    City       577 non-null    object
 5    State      577 non-null    object
 6    Chip Time  577 non-null    object
 7    Chip Pace  577 non-null    object
 8    Gun Time   577 non-null    object
 9    Team]      577 non-null    object
dtypes: object(10)
memory usage: 49.9+ KB


(581, 10)

We can perform more data cleaning by renaming the '[Place' and ' Team]' columns. 

In [ ]:
df6.rename(columns={'[Place': 'Place'},inplace=True)
df6.rename(columns={' Team]': 'Team'},inplace=True)
df6.head()

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Place    Bib  ...  Gun Time  Team
1  Finishers:   577]  ...      None  None
2       Male:   414]  ...      None  None
3     Female:   163]  ...      None  None
4           ]   None  ...      None  None
5           1    814  ...     36:24     ]

[5 rows x 10 columns]

The final data cleaning step involves removing the closing bracket for cells in the "Team" column.

In [ ]:
df6['Team'] = df6['Team'].str.strip(']')
df6.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Place    Bib  ...  Gun Time  Team
1  Finishers:   577]  ...      None  None
2       Male:   414]  ...      None  None
3     Female:   163]  ...      None  None
4           ]   None  ...      None  None
5           1    814  ...     36:24      

[5 rows x 10 columns]

Following similar steps, we can now scrape the data about the number of Olympic medals by country, available here: https://worldpopulationreview.com/country-rankings/olympic-medals-by-country

In [ ]:
url = "https://worldpopulationreview.com/country-rankings/olympic-medals-by-country"
html = urlopen(url)

In [ ]:
soup = BeautifulSoup(html, 'lxml')
type(soup)

bs4.BeautifulSoup

In [ ]:
rows = soup.find_all('tr')
print(rows[:10])

[<tr class="jsx-2642336383"><th class="jsx-2816426159" width="auto"><span class="jsx-2816426159"><a class="jsx-2816426159" data-field="country" style="color: black; cursor: pointer; display: inline-block;">Country  </a></span></th><th class="jsx-2816426159" width="auto"><span class="jsx-2816426159"><a class="jsx-2816426159" data-field="goldOlympicMedals" style="color: black; cursor: pointer; display: inline-block;">Gold  </a></span></th><th class="jsx-2816426159" width="auto"><span class="jsx-2816426159"><a class="jsx-2816426159" data-field="silverOlympicMedals" style="color: black; cursor: pointer; display: inline-block;">Silver  </a></span></th><th class="jsx-2816426159" width="auto"><span class="jsx-2816426159"><a class="jsx-2816426159" data-field="bronzeOlympicMedials" style="color: black; cursor: pointer; display: inline-block;">Bronze  </a></span></th><th class="jsx-2816426159" width="auto"><span class="jsx-2816426159"><a class="jsx-2816426159" data-field="totalOlympicMedals" sty

In [ ]:
for row in rows:
    row_td = row.find_all('td')
print(row_td)
type(row_td)

[<td><a href="/countries/sudan-population">Sudan</a></td>, <td>0</td>, <td>1</td>, <td>0</td>, <td>1</td>, <td>44,909,353</td>]


bs4.element.ResultSet

In [ ]:
str_cells = str(row_td)
cleantext = BeautifulSoup(str_cells, "lxml").get_text()
print(cleantext) 

[Sudan, 0, 1, 0, 1, 44,909,353]


In [ ]:
import re

list_rows = []
for row in rows:
    cells = row.find_all('td')
    str_cells = str(cells)
    clean = re.compile('<.*?>')
    clean2 = (re.sub(clean, '',str_cells))
    list_rows.append(clean2)
print(clean2)
type(clean2)

df = pd.DataFrame(list_rows)
df.head(10)

[Sudan, 0, 1, 0, 1, 44,909,353]


0
0                                                 []
1  [United States, 1127, 907, 793, 2827, 332,915,...
2   [United Kingdom, 274, 299, 310, 883, 68,207,116]
3          [Germany, 283, 282, 290, 855, 83,900,473]
4           [France, 248, 276, 316, 840, 65,426,179]
5            [Italy, 246, 214, 241, 701, 60,367,477]
6           [Sweden, 202, 216, 234, 652, 10,160,169]
7         [China, 237, 195, 176, 608, 1,444,216,107]
8          [Russia, 195, 163, 188, 546, 145,912,025]
9            [Norway, 188, 174, 158, 520, 5,465,630]

In [ ]:
df1 = df[0].str.split(',', expand=True)
df1.head(10)

0      1     2     3      4     5     6     7     8
0               []   None  None  None   None  None  None  None  None
1   [United States   1127   907   793   2827   332   915  073]  None
2  [United Kingdom    274   299   310    883    68   207  116]  None
3         [Germany    283   282   290    855    83   900  473]  None
4          [France    248   276   316    840    65   426  179]  None
5           [Italy    246   214   241    701    60   367  477]  None
6          [Sweden    202   216   234    652    10   160  169]  None
7           [China    237   195   176    608     1   444   216  107]
8          [Russia    195   163   188    546   145   912  025]  None
9          [Norway    188   174   158    520     5   465  630]  None

In [ ]:
df1[0] = df1[0].str.strip('[')
df1.head(10)

0      1     2     3      4     5     6     7     8
0               ]   None  None  None   None  None  None  None  None
1   United States   1127   907   793   2827   332   915  073]  None
2  United Kingdom    274   299   310    883    68   207  116]  None
3         Germany    283   282   290    855    83   900  473]  None
4          France    248   276   316    840    65   426  179]  None
5           Italy    246   214   241    701    60   367  477]  None
6          Sweden    202   216   234    652    10   160  169]  None
7           China    237   195   176    608     1   444   216  107]
8          Russia    195   163   188    546   145   912  025]  None
9          Norway    188   174   158    520     5   465  630]  None

In [ ]:
col_labels = soup.find_all('th')

In [ ]:
all_header = []
col_str = str(col_labels)
cleantext2 = BeautifulSoup(col_str, "lxml").get_text()
all_header.append(cleantext2)
print(all_header)

['[Country  , Gold  , Silver  , Bronze  , Total Olympic Medals  , 2021 Population  ]']


In [ ]:
df2 = pd.DataFrame(all_header)
df2.head()

0
0  [Country  , Gold  , Silver  , Bronze  , Total ...

In [ ]:
df3 = df2[0].str.split(',', expand=True)
df3.head()

0        1  ...                        4                    5
0  [Country     Gold    ...   Total Olympic Medals     2021 Population  ]

[1 rows x 6 columns]

In [ ]:
frames = [df3, df1]

df4 = pd.concat(frames)
df4.head(10)

0        1          2  ...     6     7     8
0      [Country     Gold     Silver    ...   NaN   NaN   NaN
0               ]     None       None  ...  None  None  None
1   United States     1127        907  ...   915  073]  None
2  United Kingdom      274        299  ...   207  116]  None
3         Germany      283        282  ...   900  473]  None
4          France      248        276  ...   426  179]  None
5           Italy      246        214  ...   367  477]  None
6          Sweden      202        216  ...   160  169]  None
7           China      237        195  ...   444   216  107]
8          Russia      195        163  ...   912  025]  None

[10 rows x 9 columns]

Your task:
    
    Choose a data source of your choice (Wikipedia is a good choice), and scrape and clean the data from it, following the steps similar to those outlined above.
    
    Your goal is to produce a data set consisting of at least 100 data points (rows), and with at least three features (columns). Your will want to:
    - Represent your data set in a cleaned up data frame format
    - In a paragraph or two, describe your data set.

In [ ]:
link = "https://www.news18.com/olympics/medal-tally/"
use_link = urlopen(link)
beauSoup = BeautifulSoup(use_link, 'lxml')
type(beauSoup)

bs4.BeautifulSoup

In [ ]:
title = beauSoup.title
title

<title>Rio Olympics 2016 Medal Tally: Rio Olympics Medal Score Latest News</title>

In [ ]:
beauSoup.find_all('a')
element = beauSoup.find_all('a')
for e in element:
  print(e.get('href'))

/
/olympics/
/olympics/
/olympics/schedule/
/olympics/medal-tally/
/newstopics/india-watch.html
/news/
#
/olympics/medal-tally/
http://www.facebook.com/sharer.php?u=https://www.news18.com/olympics/medal-tally/&t=Rio Olympics 2016 Medal Tally: Rio Olympics Medal Score Latest News
http://twitter.com/share?text=Rio Olympics 2016 Medal Tally: Rio Olympics Medal Score Latest News&url=https://www.news18.com/olympics/medal-tally/
/olympics/
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
javascript:void(0);
/olympics/schedule/5/31/
/olympics/schedule/19/31/
/olympics/schedule/6/31/
/olympics/schedule/7/31/
/olympics/schedule/9/31/
/olympics/schedule/35/31/
/olympics/schedule/10/31/
/olympics/schedule/12/31/
/olympics/schedule

In [ ]:
rows = beauSoup.find_all('tr')
print(rows[:10])

[<tr><th>Country</th> <th><img alt="" src="https://images.news18.com/static_news18/pix/ibnhome/news18/olympics-2016/images/golod-madel.png"/> Gold</th> <th><img alt="" src="https://images.news18.com/static_news18/pix/ibnhome/news18/olympics-2016/images/slidver-madel.png"/> Silver</th>
<th><img alt="" src="https://images.news18.com/static_news18/pix/ibnhome/news18/olympics-2016/images/bronz-madel.png"/> Bronze</th> <th>Total</th></tr>, <tr><td>United States(USA)</td> <td>46</td> <td>37</td> <td>38</td> <td>121</td> </tr>, <tr><td>Great Britain(GBR)</td> <td>27</td> <td>23</td> <td>17</td> <td>67</td> </tr>, <tr><td>China(CHN)</td> <td>26</td> <td>18</td> <td>26</td> <td>70</td> </tr>, <tr><td>Russia(RUS)</td> <td>19</td> <td>18</td> <td>19</td> <td>56</td> </tr>, <tr><td>Germany(GER)</td> <td>17</td> <td>10</td> <td>15</td> <td>42</td> </tr>, <tr><td>Japan(JPN)</td> <td>12</td> <td>8</td> <td>21</td> <td>41</td> </tr>, <tr><td>France(FRA)</td> <td>10</td> <td>18</td> <td>14</td> <td>42<

In [ ]:
for r in rows:
  row_tidy = r.find_all('td')
print(row_tidy)

[<td>Zimbabwe(ZIM)</td>, <td>0</td>, <td>0</td>, <td>0</td>, <td>0</td>]


In [ ]:
string = str(row_tidy)
clean = BeautifulSoup(string, 'lxml').get_text()
print(clean)

[Zimbabwe(ZIM), 0, 0, 0, 0]


In [ ]:
import re

lst_row = []
for r in rows:
  cells = r.find_all('td')
  string_cell = str(cells)
  clean = re.compile('<.*?>')
  clean2 = (re.sub(clean, '', string_cell))
  lst_row.append(clean2)
print(clean2)

df = pd.DataFrame(lst_row)
df.head(10)

[Zimbabwe(ZIM), 0, 0, 0, 0]


0
0                                     []
1  [United States(USA), 46, 37, 38, 121]
2   [Great Britain(GBR), 27, 23, 17, 67]
3           [China(CHN), 26, 18, 26, 70]
4          [Russia(RUS), 19, 18, 19, 56]
5         [Germany(GER), 17, 10, 15, 42]
6            [Japan(JPN), 12, 8, 21, 41]
7          [France(FRA), 10, 18, 14, 42]
8        [South Korea(KOR), 9, 3, 9, 21]
9             [Italy(ITA), 8, 12, 8, 28]

In [ ]:
df1 = df[0].str.split(',', expand=True)
df1.head(10)

0     1     2     3      4
0                   []  None  None  None   None
1  [United States(USA)    46    37    38   121]
2  [Great Britain(GBR)    27    23    17    67]
3          [China(CHN)    26    18    26    70]
4         [Russia(RUS)    19    18    19    56]
5        [Germany(GER)    17    10    15    42]
6          [Japan(JPN)    12     8    21    41]
7         [France(FRA)    10    18    14    42]
8    [South Korea(KOR)     9     3     9    21]
9          [Italy(ITA)     8    12     8    28]

In [ ]:
df1[0] = df1[0].str.strip('[')
df1.head(10)

0     1     2     3      4
0                   ]  None  None  None   None
1  United States(USA)    46    37    38   121]
2  Great Britain(GBR)    27    23    17    67]
3          China(CHN)    26    18    26    70]
4         Russia(RUS)    19    18    19    56]
5        Germany(GER)    17    10    15    42]
6          Japan(JPN)    12     8    21    41]
7         France(FRA)    10    18    14    42]
8    South Korea(KOR)     9     3     9    21]
9          Italy(ITA)     8    12     8    28]

In [ ]:
tb_header = beauSoup.find_all('th')
header = []
col_string = str(tb_header)
clean3 = BeautifulSoup(col_string, 'lxml').get_text()
header.append(clean3)

df2 = pd.DataFrame(header)
df2.head()

0
0  [Country,  Gold,  Silver,  Bronze, Total]

In [ ]:
df3 = df2[0].str.split(',', expand=True)
df3.head(10)

0       1         2         3        4
0  [Country    Gold    Silver    Bronze   Total]

In [ ]:
df_melt = [df3, df1]

df4 = pd.concat(df_melt)
df4.head(10)

0       1         2         3        4
0            [Country    Gold    Silver    Bronze   Total]
0                   ]    None      None      None     None
1  United States(USA)      46        37        38     121]
2  Great Britain(GBR)      27        23        17      67]
3          China(CHN)      26        18        26      70]
4         Russia(RUS)      19        18        19      56]
5        Germany(GER)      17        10        15      42]
6          Japan(JPN)      12         8        21      41]
7         France(FRA)      10        18        14      42]
8    South Korea(KOR)       9         3         9      21]

In [ ]:
df5 = df4.rename(columns = df4.iloc[0])
df5.head(5)

[Country    Gold    Silver    Bronze   Total]
0            [Country    Gold    Silver    Bronze   Total]
0                   ]    None      None      None     None
1  United States(USA)      46        37        38     121]
2  Great Britain(GBR)      27        23        17      67]
3          China(CHN)      26        18        26      70]

In [ ]:
df6 = df5.drop(df5.index[0])
df6.info()
df6.shape

<class 'pandas.core.frame.DataFrame'>
Int64Index: 208 entries, 1 to 208
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   [Country  208 non-null    object
 1     Gold    208 non-null    object
 2     Silver  208 non-null    object
 3     Bronze  208 non-null    object
 4    Total]   208 non-null    object
dtypes: object(5)
memory usage: 9.8+ KB


(208, 5)

In [ ]:
#Rename all column header titles
df6.rename(columns = {'[Country' : 'Country'}, inplace=True)
df6.rename(columns = {' Total]' : 'Total Metal'}, inplace=True)
df6.rename(columns = {'  Gold' : 'Gold'}, inplace=True)
df6.rename(columns = {'  Silver' : 'Silver'}, inplace=True)
df6.rename(columns = {'  Bronze' : 'Bronze'}, inplace=True)
df6.head()

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Country Gold Silver Bronze Total Metal
1  United States(USA)   46     37     38        121]
2  Great Britain(GBR)   27     23     17         67]
3          China(CHN)   26     18     26         70]
4         Russia(RUS)   19     18     19         56]
5        Germany(GER)   17     10     15         42]

In [ ]:
df6.rename(columns = {' Total]' : 'Total Metal'}, inplace=True)
df6['Total Metal'] = df6['Total Metal'].str.strip(']')
df6.head(20)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Country Gold Silver Bronze Total Metal
1   United States(USA)   46     37     38         121
2   Great Britain(GBR)   27     23     17          67
3           China(CHN)   26     18     26          70
4          Russia(RUS)   19     18     19          56
5         Germany(GER)   17     10     15          42
6           Japan(JPN)   12      8     21          41
7          France(FRA)   10     18     14          42
8     South Korea(KOR)    9      3      9          21
9           Italy(ITA)    8     12      8          28
10      Australia(AUS)    8     11     10          29
11    Netherlands(NED)    8      7      4          19
12        Hungary(HUN)    8      3      4          15
13         Brazil(BRA)    7      6      6          19
14          Spain(ESP)    7      4      6          17
15          Kenya(KEN)    6      6      1          13
16        Jamaica(JAM)    6      3      2          11
17        Croatia(CRO)    5      3      2          10
18           Cuba(CUB)    5      2      4          11
19    New Zealand(NZL)    4      9      5          18
20         Canada(CAN)    4      3     15          22

In [ ]:
#Counting the dataset rows
df6.shape[0]

208

In [ ]:
#Testing dataframe with inlcusion of dataset title
testing = df6
testing.style.set_caption("Olympic Games 2016 Medals Tally")

Source: [How to add table title in python preferably with pandas](https://stackoverflow.com/questions/57958432/how-to-add-table-title-in-python-preferably-with-pandas)

**Dataset Explanation**

Using a data table published through News18, the publish data table consisted of the 207 countries that attended Rio for the 2016 olympic game. The table show the total of medal each countires won and how many Gold, Silver, and Bronze metal were each countires won from the event. For instance, the United Sttes won 121 metals in total in Rio, where 46 out of 121 were golds, 47 were silvers, and 38 were bronzes. By following the provided examples, I able to scrape the data table from the news site and clean up the dataset for the lab.  The table have 208 rows — 207 countries and a row represenheader, and 4 columns — Country, Gold, Silver, Bronze, and the Total Medal. 
